fw/ss_skb.c

```diff
@@ -1609,3 +1609,289 @@ ss_skb_add_frag(struct sk_buff *skb_head, struct sk_buff *skb, char* addr,
+		__skb_frag_ref(f);
+	}
+}
+
+/* Using @split_point transform the remaining linear portion of original @skb
+ * to a fragment of the new SKB. Take care of fragments of the @skb too.
+ * If data in the linear area is not needed, just check for available head room,
+ * so the won't be any problems pushing transport headers later.
+ */
+int
+ss_skb_linear_transform(struct sk_buff *skb, struct sk_buff **skb_head,
+			unsigned char *split_point)
+{
+	struct sk_buff *nskb;
+	struct skb_shared_info *si;
+	BUG_ON(!skb_headlen(skb));
+
+	si = skb_shinfo(skb);
+	if (!split_point) {
+		/* Usage of linear portion of SKB is not expected */
+		skb_pull(skb, skb_headlen(skb));
+	} else {
+		nskb = ss_skb_alloc(0);
+		if (!nskb)
+			goto out_err;
+
+		__skb_linear_to_frag(skb, nskb,
+				     split_point - skb->data, 0);
+
+		/* can we move all the original fragments of @skb to @nskb */
+		if (likely(si->nr_frags < MAX_SKB_FRAGS)) {
+			__skb_frags_transfer(skb, nskb, 1, si->nr_frags, 0);
+			ss_skb_queue_tail(skb_head, nskb);
+		} else {
+			/* @skb contains maximum amount of fragments already */
+			ss_skb_queue_tail(skb_head, nskb);
+
+			nskb = ss_skb_alloc(0);
+			if (!nskb) {
+				__kfree_skb(*skb_head);
+				goto out_err;
+			}
+			__skb_frags_transfer(skb, nskb, 0, si->nr_frags, 0);
+			ss_skb_queue_tail(skb_head, nskb);
+		}
+	}
+	return 0;
+
+out_err:
+	T_WARN("%s: can't allocate new skb for response"
+	       " transformation(orig skb %pK)\n",
+	       __func__, skb);
+	return -ENOMEM;
+}
+
+/*
+ * This is basically a stripped-down version of skb_copy_header(),
+ * which only copies the needed SKB fields during transformation phase
+ * from below.
+ */
+static void
+__ss_skb_copy_header(struct sk_buff *oskb, struct sk_buff *nskb)
+{
+	memcpy(nskb->cb, oskb->cb, sizeof(oskb->cb));
+	nskb->dev = oskb->dev;
+
+	skb_shinfo(nskb)->gso_size = skb_shinfo(oskb)->gso_size;
+	skb_shinfo(nskb)->gso_segs = skb_shinfo(oskb)->gso_segs;
+	skb_shinfo(nskb)->gso_type = skb_shinfo(oskb)->gso_type;
+	skb_shinfo(nskb)->tx_flags = skb_shinfo(oskb)->tx_flags;
+}
+
+/*
+ * During low-headroom SKB transformation the @truesize of the replacement SKB
+ * might differ, so @sk->sk_wmem_queued and @sk->sk_forward_alloc
+ * should be adjusted accordingly.
+ */
+static void
+ss_sk_trans_mem_account(struct sock *sk, struct sk_buff *oskb,
+			  struct sk_buff *nskb)
+{
+	int delta_ts = nskb->truesize - oskb->truesize;
+	sk_wmem_queued_add(sk, delta_ts);
+	if (delta_ts > 0)
+		sk_mem_charge(sk, delta_ts);
+	else
+		sk_mem_uncharge(sk, -delta_ts);
+}
+
+/**
+ * Copy of `tcp_set_skb_tso_segs`.
+ */
+void
+ss_set_skb_tso_segs(struct sk_buff *skb, unsigned int mss_now)
+{
+	if (skb->len <= mss_now) {
+		/* Avoid the costly divide in the normal
+		 * non-TSO case.
+		 */
+		tcp_skb_pcount_set(skb, 1);
+		TCP_SKB_CB(skb)->tcp_gso_size = 0;
+	} else {
+		tcp_skb_pcount_set(skb, DIV_ROUND_UP(skb->len, mss_now));
+		TCP_SKB_CB(skb)->tcp_gso_size = mss_now;
+	}
+}
```

Comment:
We discussed this in another recent pull request - it's better to just export a Linux function instead of copying it